### PR TITLE
Register classes in ThingIFAPIBuilder

### PR DESCRIPTION
**Pull request by yongpingchen**
### New APIs
- Register Action classes
- Register ActionsResult classes
- Register TargetState classes

### Out of scope
The actual implementation to use these classes will be done during implementation phase.

### CODE REVIEW

**Comment by shinsuke-nara** [comment]
@yongpingchen 

Does `ThingIFAPI`always require the registering classes? If so, We should register it with `Builder`. What do you think?

**Comment by yongpingchen** [comment]
@shinsuke-nara these classes are not always updated. These methods should be in builder. I will fix it.

**Comment by yongpingchen** [comment]
@shinsuke-nara  fixed by your comments. Please confirm

File: thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java [hunk 1]
```diff
@@ -22,13 +25,24 @@
     private String tag;
     private final List<Schema> schemas = new ArrayList<Schema>();
 
+    private final Map<T, List<Class<? extends Action>>> actionTypes;
+    private final Map<T, List<Class<? extends ActionResult>>> actionResultTypes;
+    private final Map<T, Class<? extends TargetState>> stateTypes;
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java]: Could you add Nullable or NonNull?

File: thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java [hunk 2]
```diff
@@ -139,7 +159,46 @@ public static ThingIFAPIBuilder _newBuilder(
...
+
+    /**
+     * Register list of Action subclasses to specified alias. The registered action classes
+     * will be used for serialization/deserialization the action.
+     * If the same alias already registered, then will be updated
+     * @param alias Alias to register
+     * @param actionClasses List of Action subclasses
+     * @return builder instance for chaining call.
+     */
+    public ThingIFAPIBuilder<T> registerActions(T alias, List<Class<? extends Action>> actionClasses){
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java]: Please add Nullable or NonNull annotations to return value and an argument.

File: thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java [hunk 3]
```diff
@@ -139,7 +159,46 @@ public static ThingIFAPIBuilder _newBuilder(
...
+
+    /**
+     * Register list of ActionResult subclasses to specified alias. The registered action result classes
+     * will be used for serialization/deserialization the action result.
+     * If the same alias already registered, then will be updated
+     * @param alias Alias to register
+     * @param actionResultClasses List of ActionResult subclasses
+     * @return builder instance for chaining call.
+     */
+    public ThingIFAPIBuilder<T> registerActionResults(T alias, List<Class<? extends ActionResult>> actionResultClasses) {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java]: Please add Nullable or NonNull annotations to return value and an argument.

File: thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java [hunk 4]
```diff
@@ -139,7 +159,46 @@ public static ThingIFAPIBuilder _newBuilder(
...
+
+    /**
+     * Register TargetState to specified alias.
+     * The registered stateClass will be used when deserialization state from server.
+     * If the same alias already registered, then will be updated.
+     * @param alias Alias to register.
+     * @param stateClass Class of TargetState subclass.
+     * @return builder instance for chaining call.
+     */
+    public ThingIFAPIBuilder<T> registerTargetState(T alias, Class<TargetState> stateClass) {
```
Comment (shinsuke-nara) [inline thingif/src/main/java/com/kii/thingif/ThingIFAPIBuilder.java]: Please add Nullable or NonNull annotations to return value and an argument.

**Comment by shinsuke-nara** [comment]
@yongpingchen Could you check my comments?

**Comment by yongpingchen** [comment]
@shinsuke-nara Added nullable/nonnullable annotations, please confirm

**Comment by shinsuke-nara** [comment]
@yongpingchen LGTM!

**Pull request merged**